Monika After Story/game/mod_assets/connect4/connect4.rpy

```diff
@@ -0,0 +1,655 @@
+            self.win_tiles = set()
+            
+            # init col buttons
+            self._col_buttons = list()
+            for i in range(self.CEL_COL):
+                button = MASButtonDisplayable(
+                    Null(),
+                    Null(),
+                    Null(),
+                    Frame(self.COL_BUTTON_IDLE_IMAGE, Borders(0, 0, 0, 0)),
+                    Frame(self.COL_BUTTON_HOVER_IMAGE, Borders(0, 0, 0, 0)),
+                    Frame(self.COL_BUTTON_INSENSITIVE_IMAGE, Borders(0, 0, 0, 0)),
+                    MASConnect4Displayable.COL_BUTTON_X + MASConnect4Displayable.COL_BUTTON_X_DISTANCE * (i - 3),
+                    MASConnect4Displayable.COL_BUTTON_Y,
+                    MASConnect4Displayable.COL_BUTTON_WIDTH,
+                    MASConnect4Displayable.COL_BUTTON_HEIGHT,
+                    hover_sound=gui.hover_sound,
+                    activate_sound=gui.activate_sound,
+                    return_value=i
+                )
+                self._col_buttons.append(button)
+                
+            # init menu buttons
+            self._button_quit = MASButtonDisplayable.create_stb(
+                _("Quit"),
+                True,
+                MASConnect4Displayable.BUTTON_X,
+                MASConnect4Displayable.BUTTON_Y,
+                MASConnect4Displayable.BUTTON_WIDTH,
+                MASConnect4Displayable.BUTTON_HEIGHT,
+                hover_sound=gui.hover_sound,
+                activate_sound=gui.activate_sound
+            )
+                
+            self._button_done = MASButtonDisplayable.create_stb(
+                _("Done"),
+                False,
+                MASConnect4Displayable.BUTTON_X,
+                MASConnect4Displayable.BUTTON_Y,
+                MASConnect4Displayable.BUTTON_WIDTH,
+                MASConnect4Displayable.BUTTON_HEIGHT,
+                hover_sound=gui.hover_sound,
+                activate_sound=gui.activate_sound
+            )
+            
+            self._visible_buttons = [
+                self._button_quit
+            ]
+            
+            self._visible_buttons_winner = [
+                self._button_done
+            ]
+        
+        def render(self, width, height, st, at):
+            """
+            This function shows the board into the screen
+            """
+            render = renpy.Render(width, height)
+            
+            # prepare the board as rendererr
+            board = renpy.render(MASConnect4Displayable.BOARD_IMAGE, 1280, 720, st , at)
+            
+            # prepare winning tile highlight as renderer
+            win_tile = renpy.render(MASConnect4Displayable.WIN_TILE_IMAGE, 1280, 720, st , at)
+            
+            # prepare the pins as renderer
+            ai_pin = renpy.render(MASConnect4Displayable.AI_IMAGE, 1280, 720, st , at)
+            player_pin = renpy.render(MASConnect4Displayable.PLAYER_IMAGE, 1280, 720, st , at)
+            
```

Comment:
since it looks like these renders are only used in certain circumstances (like the game is over or 1 player is active), these should be moved to the lines where they are checked before being blit. Renders are more expensive than branch checks so the least we can do of these, the better.

Reply:
I didn't know that. I'll work on it.